src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -0,0 +1,226 @@
+section \<open>Chain Synchronization Mini-Protocol\<close>
+
+text \<open>
+  The theory in this section implements the chain synchronization mini-protocol.
+\<close>
+
+theory "Ouroboros-Mini_Protocols-Chain_Sync"
+  imports
+    "Ouroboros-Mini_Protocols"
+    "HOL-Library.BNF_Corec"
+begin
+
+subsection \<open>Cryptographic Primitives\<close>
+
+typedecl hash
+
+axiomatization
+where
+  hash_linear_order: "OFCLASS(hash, linorder_class)"
+
+instance hash :: linorder
+  by (fact hash_linear_order)
+
+subsection \<open>Blocks\<close>
+
+datatype block = Block (hash: \<open>hash\<close>)
+
+subsection \<open>Chains\<close>
+
+type_synonym chain = "block list"
+
+type_synonym chain_index = nat
+
+abbreviation (input) tip_index :: "chain \<Rightarrow> chain_index" where
+  "tip_index \<C> \<equiv> length \<C> - 1"
+
+abbreviation (input) tip :: "chain \<Rightarrow> block" where
+  "tip \<C> \<equiv> \<C> ! (tip_index \<C>)"
+
+subsection \<open>Points\<close>
+
+type_synonym point = hash
+
+definition block_point :: "block \<Rightarrow> point" where
+  [simp]: "block_point \<equiv> hash"
+
+subsection \<open>Protocol Parameters\<close>
+
+datatype phase =
+  is_intersection_finding: IntersectionFinding |
+  is_chain_update: ChainUpdate
+
+type_synonym consumer_parameters = "
+  chain \<comment> \<open>current chain\<close> \<times>
+  phase \<comment> \<open>current phase\<close> \<times>
+  (chain \<Rightarrow> point list)" \<comment> \<open>function for selecting candidate points\<close>
+
+type_synonym producer_parameters = "
+  chain \<comment> \<open>current chain\<close> \<times>
+  chain_index \<comment> \<open>read pointer\<close> \<times>
+  bool \<comment> \<open>whether the next response to \<open>RequestNext\<close> should be \<open>RollBackward\<close>\<close>"
+
+locale chain_sync =
+  fixes initial_params\<^sub>c :: consumer_parameters
+    and initial_params\<^sub>p :: producer_parameters
+
+subsection \<open>Parties\<close>
+
+datatype party =
+  Producer |
+  Consumer
```

Comment:
Is there any particular reason for you having chosen the terms “producer” and “consumer” instead of “client” and “server”, as I have done for the dummy protocols?

Sadly, *The Shelley Networking Protocol* doesn’t use a single scheme for naming parties but seemingly randomly mixes three such schemes: client–server, producer–consumer, and initiator–responder. What’s worse, the correspondence between names from different naming schemes seems to depend on the protocol: for example, the client is usually the consumer, but in the case of the transaction submission protocol it is apparently the producer.

I think we should settle on a single naming scheme for all our mini-protocol specifications. Client–server seems appropriate to me, since the terms “client” and “server” seem to correspond to the same colors used in state machines, independently of the mini-protocol (although the colors are swapped for the local transaction submission protocol, which seems to be a bug and which I’ve already reported to @bolt12). Also, I don’t like the term producer, since the so-called producers typically don’t produce: with chain synchronization, for example, the “producer” normally doesn’t produce the whole chain. “Provider” would have been a better name.